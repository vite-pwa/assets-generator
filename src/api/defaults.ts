import type { PngOptions, ResizeOptions } from 'sharp'
import type { AssetType } from '../types.ts'

export const defaultResizeOptions = {
  fit: 'contain',
  background: 'white',
} satisfies ResizeOptions

export const defaultDarkResizeOptions = {
  fit: 'contain',
  background: 'black',
} satisfies ResizeOptions

export const defaultPngCompressionOptions: PngOptions = {
  compressionLevel: 9,
  quality: 60,
}

export const defaultPngOptions: Record<AssetType, { padding: number, resizeOptions: ResizeOptions }> = {
  transparent: { padding: 0.05, resizeOptions: { fit: 'contain', background: 'transparent' } },
  maskable: { padding: 0.3, resizeOptions: { fit: 'contain', background: 'white' } },
  apple: { padding: 0.3, resizeOptions: { fit: 'contain', background: 'white' } },
}

export function createResizeOptions(dark: boolean, options: ResizeOptions) {
  return {
    ...(dark ? defaultDarkResizeOptions : defaultResizeOptions),
    ...options,
  } satisfies ResizeOptions
}

export function createPngCompressionOptions(options: PngOptions) {
  return {
    ...defaultPngCompressionOptions,
    ...options,
  } satisfies PngOptions
}
