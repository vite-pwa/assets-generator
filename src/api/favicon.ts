import type { GenerateFaviconOptionsType, GenerateFaviconType, ImageSourceInput } from './types.ts'
import sharp from 'sharp'
import { encode } from 'sharp-ico'

export async function generateFavicon<Format extends GenerateFaviconType>(
  format: Format,
  image: ImageSourceInput,
  options?: GenerateFaviconOptionsType<Format>,
) {
  return encode([await sharp(image).toFormat(format, options).toBuffer()])
}

if (import.meta.vitest) {
  const { expect, it } = import.meta.vitest
  it('should generate a favicon asset', async () => {
    const result = await generateFavicon(
      'png',
      'playground/pwa/public/favicon.svg',
    )
    expect(result).toBeDefined()
  })
}
