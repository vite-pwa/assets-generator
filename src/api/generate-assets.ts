import type { IconAsset, ImageAssetsInstructions } from './types.ts'
import { existsSync } from 'node:fs'
import { writeFile } from 'node:fs/promises'
import { resolve } from 'node:path'

export async function generateAssets(
  instruction: ImageAssetsInstructions,
  overrideAssets: boolean,
  folder: string,
  log?: (message: string, ignored: boolean) => void,
) {
  const transparent = Array.from(Object.values(instruction.transparent))
  await Promise.all(transparent.map(icon => generateAsset('PNG', icon, folder, overrideAssets, log)))
  const maskable = Array.from(Object.values(instruction.maskable))
  await Promise.all(maskable.map(icon => generateAsset('PNG', icon, folder, overrideAssets, log)))
  const apple = Array.from(Object.values(instruction.apple))
  await Promise.all(apple.map(icon => generateAsset('PNG', icon, folder, overrideAssets, log)))
  const favicon = Array.from(Object.values(instruction.favicon))
  await Promise.all(favicon.filter(icon => !icon.name.endsWith('.svg')).map(icon => generateAsset('ICO', icon, folder, overrideAssets, log)))
  const appleSplashScreen = Array.from(Object.values(instruction.appleSplashScreen))
  await Promise.all(appleSplashScreen.map(icon => generateAsset('PNG', icon, folder, overrideAssets, log)))
}

async function generateAsset(
  type: 'ICO' | 'PNG',
  icon: IconAsset<any>,
  folder: string,
  overrideAssets: boolean,
  log?: (message: string, ignored: boolean) => void,
) {
  const filePath = resolve(folder, icon.name)
  if (!overrideAssets && existsSync(filePath)) {
    log?.(`Skipping, ${type} file already exists: ${filePath}`, true)
    return
  }

  await icon
    .buffer()
    .then(b => writeFile(resolve(folder, icon.name), new Uint8Array(b)))
    .then(() => {})

  log?.(`Generated ${type} file: ${filePath}`, false)
}
