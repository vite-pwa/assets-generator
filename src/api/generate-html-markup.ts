import type { ImageAssetsInstructions } from './types.ts'

export function generateHtmlMarkup(instruction: ImageAssetsInstructions) {
  const apple = Array.from(Object.values(instruction.apple))
  const favicon = Array.from(Object.values(instruction.favicon))
  const appleSplashScreen = Array.from(Object.values(instruction.appleSplashScreen))
  const links: string[] = []
  favicon.forEach(icon => icon.link?.length && links.push(icon.link))
  apple.forEach(icon => icon.link?.length && links.push(icon.link))
  appleSplashScreen.forEach(icon => icon.link?.length && links.push(icon.link))

  return links
}
