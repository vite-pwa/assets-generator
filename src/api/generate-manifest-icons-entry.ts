import type { ImageAssetsInstructions, ManifestIcons, ManifestIconsOptionsType, ManifestIconsType } from './types.ts'

export function generateManifestIconsEntry<Format extends ManifestIconsType>(
  format: Format,
  instruction: ImageAssetsInstructions,
): ManifestIconsOptionsType<Format> {
  const icons: ManifestIcons = { icons: [] }

  for (const icon of Object.values(instruction.transparent)) {
    icons.icons.push({
      src: icon.url,
      sizes: `${icon.width}x${icon.height}`,
      type: icon.mimeType,
    })
  }
  for (const icon of Object.values(instruction.maskable)) {
    icons.icons.push({
      src: icon.url,
      sizes: `${icon.width}x${icon.height}`,
      type: icon.mimeType,
      purpose: 'maskable',
    })
  }

  return (
    format === 'string'
      ? JSON.stringify(icons, null, 2)
      : icons
  ) as ManifestIconsOptionsType<Format>
}
