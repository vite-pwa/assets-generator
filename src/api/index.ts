export * from './types.ts'
// eslint-disable-next-line perfectionist/sort-exports
export * from './defaults.ts'
export * from './utils.ts'
// eslint-disable-next-line perfectionist/sort-exports
export * from './html.ts'
export * from './maskable.ts'
export * from './transparent.ts'
// eslint-disable-next-line perfectionist/sort-exports
export * from './favicon.ts'
