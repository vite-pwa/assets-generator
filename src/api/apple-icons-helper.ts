import type { PngOptions, ResizeOptions } from 'sharp'
import type { AppleDeviceSize, AppleSplashScreens, ResolvedAppleSplashScreens } from '../types.ts'
import type { ImageAssets, ImageAssetsInstructions } from './types.ts'
import { defaultSplashScreenName } from '../splash.ts'
import { createPngCompressionOptions, createResizeOptions } from './defaults.ts'
import { createAppleSplashScreenHtmlLink } from './html.ts'
import { generateMaskableAsset } from './maskable.ts'

export function resolveAppleSplashScreensInstructions(
  // eslint-disable-next-line node/prefer-global/buffer
  image: Buffer,
  imageAssets: ImageAssets,
  instructions: ImageAssetsInstructions,
  useAppleSplashScreens?: AppleSplashScreens,
) {
  const appleSplashScreens = resolveAppleSplashScreens(
    imageAssets,
    useAppleSplashScreens,
  )
  if (!appleSplashScreens || !appleSplashScreens.sizes.length)
    return

  const { linkMediaOptions, name: resolveName, sizes, png } = appleSplashScreens

  const sizesMap = new Map<number, number>()
  const splashScreens: SplashScreenData[] = sizes.reduce((acc, size) => {
    // cleanup duplicates screen dimensions:
    // should we add the links (maybe scaleFactor is different)?
    if (sizesMap.get(size.width) === size.height)
      return acc

    sizesMap.set(size.width, size.height)
    const { width: height, height: width, ...restSize } = size
    const {
      width: lheight,
      height: lwidth,
      ...restResizeOptions
    } = size.resizeOptions || {}
    const landscapeSize: AppleDeviceSize = {
      ...restSize,
      width,
      height,
      resizeOptions: {
        ...restResizeOptions,
        width: lwidth,
        height: lheight,
      },
    }
    acc.push({
      size,
      landscape: false,
      dark: size.darkResizeOptions ? false : undefined,
      resizeOptions: size.resizeOptions,
      padding: size.padding ?? 0.3,
      png: size.png ?? png,
    })
    acc.push({
      size: landscapeSize,
      landscape: true,
      dark: size.darkResizeOptions ? false : undefined,
      resizeOptions: landscapeSize.resizeOptions,
      padding: size.padding ?? 0.3,
      png: size.png ?? png,
    })
    if (size.darkResizeOptions) {
      const {
        width: dlheight,
        height: dlwidth,
        ...restDarkResizeOptions
      } = size.darkResizeOptions
      const landscapeDarkResizeOptions: ResizeOptions = { ...restDarkResizeOptions, width: dlwidth, height: dlheight }
      const landscapeDarkSize: AppleDeviceSize = {
        ...restSize,
        width,
        height,
        resizeOptions: landscapeDarkResizeOptions,
        darkResizeOptions: undefined,
      }
      acc.push({
        size,
        landscape: false,
        dark: true,
        resizeOptions: size.darkResizeOptions,
        padding: size.padding ?? 0.3,
        png: size.png ?? png,
      })
      acc.push({
        size: landscapeDarkSize,
        landscape: true,
        dark: true,
        resizeOptions: landscapeDarkResizeOptions,
        padding: size.padding ?? 0.3,
        png: size.png ?? png,
      })
    }
    return acc
  }, [] as SplashScreenData[])

  sizesMap.clear()

  // eslint-disable-next-line node/prefer-global/buffer
  const cache: Record<string, Promise<Buffer>> = {}
  const originalName = imageAssets.originalName!

  // eslint-disable-next-line node/prefer-global/buffer
  const imageResolver = (dark: boolean): Promise<Buffer> => {
    if (!dark || typeof appleSplashScreens.darkImageResolver !== 'function')
      return Promise.resolve(image)

    const cached = cache[originalName]
    if (cached)
      return cached

    return cache[originalName] = appleSplashScreens
      .darkImageResolver(originalName)
      .then(darkImage => Promise.resolve(darkImage ?? image))
  }

  for (const size of splashScreens) {
    const basePath = linkMediaOptions.basePath
    const name = resolveName(size.landscape, size.size, size.dark)
    const url = `${basePath}${name}`
    const promise = () => imageResolver(size.dark === true).then(i => generateMaskableAsset('png', i, size.size, {
      padding: size.padding,
      resizeOptions: {
        ...size.resizeOptions,
        background: size.resizeOptions?.background ?? (size.dark ? 'black' : 'white'),
      },
      outputOptions: size.png,
    }))

    instructions.appleSplashScreen[url] = {
      name,
      url,
      width: size.size.width,
      height: size.size.height,
      mimeType: 'image/png',
      link: createAppleSplashScreenHtmlLink('string', {
        size: size.size,
        landscape: size.landscape,
        addMediaScreen: linkMediaOptions.addMediaScreen,
        xhtml: linkMediaOptions.xhtml,
        name: resolveName,
        basePath,
        dark: size.dark,
        includeId: imageAssets.htmlLinks.includeId,
      }),
      linkObject: createAppleSplashScreenHtmlLink('link', {
        size: size.size,
        landscape: size.landscape,
        addMediaScreen: linkMediaOptions.addMediaScreen,
        xhtml: linkMediaOptions.xhtml,
        name: resolveName,
        basePath,
        dark: size.dark,
        includeId: imageAssets.htmlLinks.includeId,
      }),
      buffer: () => promise().then(m => m.toBuffer()),
    }
  }
}

interface SplashScreenData {
  size: AppleDeviceSize
  landscape: boolean
  dark?: boolean
  resizeOptions?: ResizeOptions
  padding: number
  png: PngOptions
}

function resolveAppleSplashScreens(
  imageAssets: ImageAssets,
  useAppleSplashScreens?: AppleSplashScreens,
) {
  let appleSplashScreens: ResolvedAppleSplashScreens | undefined
  if (useAppleSplashScreens) {
    const {
      padding = 0.3,
      resizeOptions: useResizeOptions = {},
      darkResizeOptions: useDarkResizeOptions,
      linkMediaOptions: useLinkMediaOptions = {},
      sizes,
      name = defaultSplashScreenName,
      png: usePng = {},
      darkImageResolver,
    } = useAppleSplashScreens

    // Initialize defaults
    const resizeOptions = createResizeOptions(false, useResizeOptions)
    const darkResizeOptions = useDarkResizeOptions ? createResizeOptions(true, useDarkResizeOptions) : undefined
    const png: PngOptions = createPngCompressionOptions(usePng)

    for (const size of sizes) {
      if (typeof size.padding === 'undefined')
        size.padding = padding

      if (typeof size.png === 'undefined')
        size.png = png

      if (typeof size.resizeOptions === 'undefined')
        size.resizeOptions = resizeOptions

      if (typeof size.darkResizeOptions === 'undefined' && darkResizeOptions)
        size.darkResizeOptions = darkResizeOptions
    }
    const {
      log = true,
      addMediaScreen = true,
      basePath = imageAssets.basePath ?? '/',
      xhtml = false,
    } = useLinkMediaOptions
    appleSplashScreens = {
      darkImageResolver,
      padding,
      sizes,
      linkMediaOptions: {
        log,
        addMediaScreen,
        basePath,
        xhtml,
      },
      name,
      png,
    }
  }

  return appleSplashScreens
}
