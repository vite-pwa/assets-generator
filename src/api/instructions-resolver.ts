import type { PngOptions, ResizeOptions } from 'sharp'
import type {
  AppleDeviceSize,
  AppleSplashScreens,
  BuiltInPreset,
  Preset,
  ResolvedAppleSplashScreens,
  ResolvedAssets,
} from '../config.ts'
import {
  createPngCompressionOptions,
  createResizeOptions,
  defaultAssetName,
  defaultPngCompressionOptions,
  defaultSplashScreenName,
} from '../config.ts'
import { toResolvedAsset, toResolvedSize } from '../utils.ts'
import type { HtmlLinkPreset } from './html.ts'
import {
  createAppleSplashScreenHtmlLink,
  createAppleTouchIconHtmlLink,
  createFaviconHtmlLink,
} from './html.ts'
import type { ImageAssets, ImageAssetsInstructions, ImageSourceInput } from './types.ts'
import { generateTransparentAsset } from './transparent.ts'
import { generateMaskableAsset } from './maskable.ts'
import { generateFavicon } from './favicon.ts'

export async function resolveInstructions(imageAssets: ImageAssets) {
  const {
    imageResolver,
    imageName,
    originalName,
    preset = 'minimal',
    faviconPreset,
  } = imageAssets

  const [usePreset, htmlPreset] = await resolvePreset(preset, faviconPreset)

  const {
    assetName = defaultAssetName,
    png = defaultPngCompressionOptions,
    appleSplashScreens: useAppleSplashScreens,
  } = usePreset

  const assets: ResolvedAssets = {
    assets: {
      transparent: toResolvedAsset('transparent', usePreset.transparent),
      maskable: toResolvedAsset('maskable', usePreset.maskable),
      apple: toResolvedAsset('apple', usePreset.apple),
    },
    png,
    assetName,
  }

  const instructions = {
    image: imageName,
    originalName,
    favicon: {},
    transparent: {},
    maskable: {},
    apple: {},
    appleSplashScreen: {},
  } as ImageAssetsInstructions

  const image = await imageResolver()

  resolveTransparentIcons(imageAssets, image, assets, htmlPreset, instructions)
  resolveMaskableIcons('maskable', imageAssets, image, assets, htmlPreset, instructions)
  resolveMaskableIcons('apple', imageAssets, image, assets, htmlPreset, instructions)

  if (imageName.endsWith('.svg')) {
    const name = imageAssets.resolveSvgName(imageName)
    const url = `${imageAssets.basePath}${name}`
    instructions.favicon[url] = {
      name,
      url,
      width: 0,
      height: 0,
      mimeType: 'image/svg+xml',
      link: createFaviconHtmlLink('string', htmlPreset, {
        name,
        basePath: imageAssets.basePath,
        xhtml: imageAssets.htmlLinks.xhtml,
        includeId: imageAssets.htmlLinks.includeId,
      }),
      linkObject: createFaviconHtmlLink('link', htmlPreset, {
        name,
        basePath: imageAssets.basePath,
        xhtml: imageAssets.htmlLinks.xhtml,
        includeId: imageAssets.htmlLinks.includeId,
      }),
      buffer: () => Promise.resolve(image),
    }
  }

  resolveAppleSplashScreensInstructions(imageAssets, instructions, useAppleSplashScreens)

  return instructions
}

async function resolvePreset(preset: BuiltInPreset | Preset, faviconPreset?: HtmlLinkPreset): Promise<[Preset, HtmlLinkPreset]> {
  if (typeof preset === 'object')
    return [preset, faviconPreset ?? 'default']

  switch (preset) {
    case 'minimal':
      return [await import('../presets/minimal.ts').then(m => m.minimalPreset), 'default']
    case 'minimal-2023':
      return [await import('../presets/minimal-2023.ts').then(m => m.minimal2023Preset), '2023']
    default:
      throw new Error(`Preset ${preset} not yet implemented`)
  }
}

function resolveTransparentIcons(
  imageAssets: ImageAssets,
  image: ImageSourceInput,
  assets: ResolvedAssets,
  htmlPreset: HtmlLinkPreset,
  instructions: ImageAssetsInstructions,
) {
  const asset = assets.assets.transparent
  const { sizes, padding, resizeOptions } = asset
  const { basePath, htmlLinks: { xhtml, includeId } } = imageAssets
  for (const size of sizes) {
    const name = assets.assetName('transparent', size)
    const url = `${basePath}${name}`
    const promise = () => generateTransparentAsset('png', image, size, {
      padding,
      resizeOptions,
      outputOptions: assets.png,
    })
    instructions.transparent[url] = {
      name,
      url,
      width: size.width,
      height: size.height,
      mimeType: 'image/png',
      buffer: () => promise().then(m => m.toBuffer()),
    }
  }

  const favicons = asset.favicons
  if (!favicons)
    return

  for (const [size, name] of favicons) {
    const url = `${basePath}${name}`
    const promise = () => generateTransparentAsset('png', image, size, {
      padding,
      resizeOptions,
      outputOptions: assets.png,
    }).then(m => m.toBuffer())
      .then(b => generateFavicon('png', b))
    const resolvedSize = toResolvedSize(size)
    instructions.favicon[url] = {
      name,
      url,
      width: resolvedSize.width,
      height: resolvedSize.height,
      mimeType: 'image/x-icon',
      link: createFaviconHtmlLink('string', htmlPreset, {
        name,
        size,
        basePath,
        xhtml,
        includeId,
      }),
      linkObject: createFaviconHtmlLink('link', htmlPreset, {
        name,
        size,
        basePath,
        xhtml,
        includeId,
      }),
      buffer: () => promise(),
    }
  }
}

function resolveMaskableIcons(
  type: 'apple' | 'maskable',
  imageAssets: ImageAssets,
  image: ImageSourceInput,
  assets: ResolvedAssets,
  htmlPreset: HtmlLinkPreset,
  instructions: ImageAssetsInstructions,
) {
  const asset = assets.assets[type]
  const { sizes, padding, resizeOptions } = asset
  const { basePath, htmlLinks: { xhtml, includeId } } = imageAssets
  for (const size of sizes) {
    const name = assets.assetName(type, size)
    const url = `${basePath}${name}`
    const promise = () => generateMaskableAsset('png', image, size, {
      padding,
      resizeOptions,
      outputOptions: assets.png,
    })
    const buffer = () => promise().then(m => m.toBuffer())
    if (type === 'apple') {
      instructions.apple[url] = {
        name,
        url,
        width: size.width,
        height: size.height,
        mimeType: 'image/png',
        link: createAppleTouchIconHtmlLink('string', {
          name,
          size,
          basePath,
          xhtml,
          includeId,
        }),
        linkObject: createAppleTouchIconHtmlLink('link', {
          name,
          size,
          basePath,
          xhtml,
          includeId,
        }),
        buffer,
      }
    }
    else {
      instructions.maskable[url] = {
        name,
        url,
        width: size.width,
        height: size.height,
        mimeType: 'image/png',
        buffer,
      }
    }
  }

  const favicons = asset.favicons
  if (!favicons)
    return

  for (const [size, name] of favicons) {
    const url = `${basePath}${name}`
    const resolvedSize = toResolvedSize(size)
    instructions.favicon[url] = {
      name,
      url,
      width: resolvedSize.width,
      height: resolvedSize.height,
      mimeType: 'image/x-icon',
      link: createFaviconHtmlLink('string', htmlPreset, {
        name,
        size,
        basePath,
        xhtml,
        includeId,
      }),
      linkObject: createFaviconHtmlLink('link', htmlPreset, {
        name,
        size,
        basePath,
        xhtml,
        includeId,
      }),
      buffer: () => generateMaskableAsset('png', image, size, {
        padding,
        resizeOptions,
        outputOptions: assets.png,
      }).then(m => m.toBuffer()).then(b => generateFavicon('png', b)),
    }
  }
}

function resolveAppleSplashScreens(
  useAppleSplashScreens?: AppleSplashScreens,
) {
  let appleSplashScreens: ResolvedAppleSplashScreens | undefined
  if (useAppleSplashScreens) {
    const {
      padding = 0.3,
      resizeOptions: useResizeOptions = {},
      darkResizeOptions: useDarkResizeOptions = {},
      linkMediaOptions: useLinkMediaOptions = {},
      sizes,
      name = defaultSplashScreenName,
      png: usePng = {},
    } = useAppleSplashScreens

    // Initialize defaults
    const resizeOptions = createResizeOptions(false, useResizeOptions)
    const darkResizeOptions = createResizeOptions(true, useDarkResizeOptions)
    const png: PngOptions = createPngCompressionOptions(usePng)

    for (const size of sizes) {
      if (typeof size.padding === 'undefined')
        size.padding = padding

      if (typeof size.png === 'undefined')
        size.png = png

      if (typeof size.resizeOptions === 'undefined')
        size.resizeOptions = resizeOptions

      if (typeof size.darkResizeOptions === 'undefined')
        size.darkResizeOptions = darkResizeOptions
    }
    const {
      log = true,
      addMediaScreen = true,
      basePath = '/',
      xhtml = false,
    } = useLinkMediaOptions
    appleSplashScreens = {
      padding,
      sizes,
      linkMediaOptions: {
        log,
        addMediaScreen,
        basePath,
        xhtml,
      },
      name,
      png,
    }
  }

  return appleSplashScreens
}

interface SplashScreenData {
  size: AppleDeviceSize
  landscape: boolean
  dark?: boolean
  resizeOptions?: ResizeOptions
  padding: number
  png: PngOptions
}

function resolveAppleSplashScreensInstructions(
  imageAssets: ImageAssets,
  instructions: ImageAssetsInstructions,
  useAppleSplashScreens?: AppleSplashScreens,
) {
  const appleSplashScreens = resolveAppleSplashScreens(useAppleSplashScreens)
  if (!appleSplashScreens || !appleSplashScreens.sizes.length)
    return

  const { linkMediaOptions, name: resolveName, sizes, png } = appleSplashScreens

  const sizesMap = new Map<number, number>()
  const splashScreens: SplashScreenData[] = sizes.reduce((acc, size) => {
    // cleanup duplicates screen dimensions:
    // should we add the links (maybe scaleFactor is different)?
    if (sizesMap.get(size.width) === size.height)
      return acc

    sizesMap.set(size.width, size.height)
    const { width: height, height: width, ...restSize } = size
    const {
      width: lheight,
      height: lwidth,
      ...restResizeOptions
    } = size.resizeOptions || {}
    const landscapeSize: AppleDeviceSize = {
      ...restSize,
      width,
      height,
      resizeOptions: {
        ...restResizeOptions,
        width: lwidth,
        height: lheight,
      },
    }
    acc.push({
      size,
      landscape: false,
      dark: size.darkResizeOptions ? false : undefined,
      resizeOptions: size.resizeOptions,
      padding: size.padding ?? 0.3,
      png: size.png ?? png,
    })
    acc.push({
      size: landscapeSize,
      landscape: true,
      dark: size.darkResizeOptions ? false : undefined,
      resizeOptions: landscapeSize.resizeOptions,
      padding: size.padding ?? 0.3,
      png: size.png ?? png,
    })
    if (size.darkResizeOptions) {
      const {
        width: dlheight,
        height: dlwidth,
        ...restDarkResizeOptions
      } = size.darkResizeOptions
      const landscapeDarkResizeOptions: ResizeOptions = { ...restDarkResizeOptions, width: dlwidth, height: dlheight }
      const landscapeDarkSize: AppleDeviceSize = {
        ...restSize,
        width,
        height,
        resizeOptions: landscapeDarkResizeOptions,
        darkResizeOptions: undefined,
      }
      acc.push({
        size,
        landscape: false,
        dark: true,
        resizeOptions: size.darkResizeOptions,
        padding: size.padding ?? 0.3,
        png: size.png ?? png,
      })
      acc.push({
        size: landscapeDarkSize,
        landscape: true,
        dark: true,
        resizeOptions: landscapeDarkResizeOptions,
        padding: size.padding ?? 0.3,
        png: size.png ?? png,
      })
    }
    return acc
  }, [] as SplashScreenData[])

  sizesMap.clear()

  for (const size of splashScreens) {
    const name = resolveName(size.landscape, size.size, size.dark)
    const url = `${imageAssets.basePath}${name}`
    const promise = () => generateMaskableAsset('png', imageAssets.imageName, size.size, {
      padding: size.padding,
      resizeOptions: {
        ...size.resizeOptions,
        background: size.resizeOptions?.background ?? (size.dark ? 'black' : 'white'),
      },
      outputOptions: size.png,
    })
    instructions.appleSplashScreen[url] = {
      name,
      url,
      width: size.size.width,
      height: size.size.height,
      mimeType: 'image/png',
      link: createAppleSplashScreenHtmlLink('string', {
        size: size.size,
        landscape: size.landscape,
        addMediaScreen: linkMediaOptions.addMediaScreen,
        xhtml: linkMediaOptions.xhtml,
        name: resolveName,
        basePath: linkMediaOptions.basePath,
        dark: size.dark,
        includeId: imageAssets.htmlLinks.includeId,
      }),
      linkObject: createAppleSplashScreenHtmlLink('link', {
        size: size.size,
        landscape: size.landscape,
        addMediaScreen: linkMediaOptions.addMediaScreen,
        xhtml: linkMediaOptions.xhtml,
        name: resolveName,
        basePath: linkMediaOptions.basePath,
        dark: size.dark,
        includeId: imageAssets.htmlLinks.includeId,
      }),
      buffer: () => promise().then(m => m.toBuffer()),
    }
  }
}
