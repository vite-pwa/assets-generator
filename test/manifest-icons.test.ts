import { readFile } from 'node:fs/promises'
import { basename } from 'node:path'
import { expect, expectTypeOf, it } from 'vitest'
import { resolveInstructions } from '../src/api/instructions-resolver'
import { generateManifestIconsEntry } from '../src/api/generate-manifest-icons-entry'
import type { ManifestIcons } from '../src/api'

it('should generate manifest icons entry', async () => {
  const instructions = await resolveInstructions({
    imageResolver: () => readFile('playground/pwa/public/favicon.svg'),
    imageName: 'playground/pwa/public/favicon.svg',
    preset: 'minimal-2023',
    htmlLinks: {
      xhtml: false,
      includeId: false,
    },
    basePath: '/',
    resolveSvgName: name => basename(name),
  })
  const iconsString = generateManifestIconsEntry('string', instructions)
  expectTypeOf(iconsString).toEqualTypeOf<string>()
  const icons = generateManifestIconsEntry('object', instructions)
  expectTypeOf(icons).toEqualTypeOf<ManifestIcons>()
  expect(icons).toMatchSnapshot()
})
