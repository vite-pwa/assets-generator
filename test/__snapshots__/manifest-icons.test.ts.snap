// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should generate manifest icons entry 1`] = `
{
  "icons": [
    {
      "sizes": "64x64",
      "src": "pwa-64x64.png",
      "type": "image/png",
    },
    {
      "sizes": "192x192",
      "src": "pwa-192x192.png",
      "type": "image/png",
    },
    {
      "sizes": "512x512",
      "src": "pwa-512x512.png",
      "type": "image/png",
    },
    {
      "purpose": "maskable",
      "sizes": "512x512",
      "src": "maskable-icon-512x512.png",
      "type": "image/png",
    },
  ],
}
`;
